src/lem/constrainer.rs

```diff
@@ -81,103 +150,565 @@ impl LEM {
         match concrete_path {
-            None => Ok(true),
-            Some(concrete_path) => concrete_path
+            PathKind::Concrete => Ok(true),
+            PathKind::MaybeConcrete(concrete_path) => concrete_path
                 .get_value()
                 .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
         }
     }
 
+    fn z_ptr_from_valuation<F: LurkField>(
+        path_kind: &PathKind,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(path_kind)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn stack_hash_slots<F: LurkField>(
+        path_kind: &PathKind,
+        hash_slots: &mut HashSlots<F>,
+        slots: &SlotsIndices,
+        hash: MetaPtr,
+        alloc_arity: AllocHashPreimage<F>,
+        tag: Option<Tag>,
+    ) -> Result<()> {
+        let is_concrete_path = Self::on_concrete_path(path_kind)?;
+        match alloc_arity {
+            AllocHashPreimage::A2(i0, i1) => {
+                if let PathKind::MaybeConcrete(concrete_path) = path_kind {
+                    // if concrete_path is true, push to slots
+                    if is_concrete_path {
+                        hash_slots.hash2_alloc.push((slots.hash2_idx, i0, i1));
+                        // only once per path
+                    }
+                    // concrete path implies alloc_tgt has the same value as in the current slot
+                    hash_slots.hash2_stacks.implies_stack.push((
+                        concrete_path.clone(),
+                        slots.hash2_idx,
+                        hash,
+                        tag,
+                    ));
+                // many
+                } else {
```

Comment:
Now that you have `PathKind`, I think using a `match` instead of `if` would make it easier to understand what this `else` implies down here.